File: ts/test/entity/invoice_extraction/InvoiceExtractionEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { InvoiceExtractionSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('InvoiceExtractionEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when INVOICEEXTRACTION_TEST_LIVE=TRUE.
  afterEach(liveDelay('INVOICEEXTRACTION_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = InvoiceExtractionSDK.test()
    const ent = testsdk.InvoiceExtraction()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.INVOICE_EXTRACTION_TEST_LIVE
    for (const op of ['create']) {
      if (maybeSkipControl(t, 'entityOp', 'invoice_extraction.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set INVOICE_EXTRACTION_TEST_INVOICE_EXTRACTION_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const invoice_extraction_ref01_ent = client.InvoiceExtraction()
    let invoice_extraction_ref01_data = setup.data.new.invoice_extraction['invoice_extraction_ref01']

    invoice_extraction_ref01_data = await invoice_extraction_ref01_ent.create(invoice_extraction_ref01_data)
    assert(null != invoice_extraction_ref01_data)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/invoice_extraction/InvoiceExtractionTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = InvoiceExtractionSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['invoice_extraction01','invoice_extraction02','invoice_extraction03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['INVOICE_EXTRACTION_TEST_INVOICE_EXTRACTION_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'INVOICE_EXTRACTION_TEST_INVOICE_EXTRACTION_ENTID': idmap,
    'INVOICE_EXTRACTION_TEST_LIVE': 'FALSE',
    'INVOICE_EXTRACTION_TEST_EXPLAIN': 'FALSE',
    'INVOICE_EXTRACTION_APIKEY': 'NONE',
  })

  idmap = env['INVOICE_EXTRACTION_TEST_INVOICE_EXTRACTION_ENTID']

  const live = 'TRUE' === env.INVOICE_EXTRACTION_TEST_LIVE

  if (live) {
    client = new InvoiceExtractionSDK(merge([
      {
        apikey: env.INVOICE_EXTRACTION_APIKEY,
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.INVOICE_EXTRACTION_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  

File: rb/test/invoice_extraction_entity_test.rb
# InvoiceExtraction entity test

require "minitest/autorun"
require "json"
require_relative "../InvoiceExtraction_sdk"
require_relative "runner"

class InvoiceExtractionEntityTest < Minitest::Test
  def test_create_instance
    testsdk = InvoiceExtractionSDK.test(nil, nil)
    ent = testsdk.InvoiceExtraction(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = invoice_extraction_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "invoice_extraction." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set INVOICE_EXTRACTION_TEST_INVOICE_EXTRACTION_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    invoice_extraction_ref01_ent = client.InvoiceExtraction(nil)
    invoice_extraction_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.invoice_extraction"), "invoice_extraction_ref01"))

    invoice_extraction_ref01_data_result = invoice_extraction_ref01_ent.create(invoice_extraction_ref01_data, nil)
    invoice_extraction_ref01_data = Helpers.to_map(invoice_extraction_ref01_data_result.respond_to?(:data_get) ? invoice_extraction_ref01_data_result.data_get : invoice_extraction_ref01_data_result)
    assert !invoice_extraction_ref01_data.nil?

  end
end

def invoice_extraction_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "invoice_extraction", "InvoiceExtractionTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = InvoiceExtractionSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["invoice_extraction01", "invoice_extraction02", "invoice_extraction03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["INVOICE_EXTRACTION_TEST_INVOICE_EXTRACTION_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "INVOICE_EXTRACTION_TEST_INVOICE_EXTRACTION_ENTID" => idmap,
    "INVOICE_EXTRACTION_TEST_LIVE" => "FALSE",
    "INVOICE_EXTRACTION_TEST_EXPLAIN" => "FALSE",
    "INVOICE_EXTRACTION_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["INVOICE_EXTRACTION_TEST_INVOICE_EXTRACTION_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["INVOICE_EXTRACTION_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["INVOICE_EXTRACTION_APIKEY"],
      },
      extra || {},
    ])
    client = InvoiceExtractionSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["INVOICE_EXTRACTION_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["INVOICE_EXTRACTION_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
